File: utils/metadata.py
"""
utils/metadata.py — Resolve Stremio IDs to title + year via Cinemeta/Kitsu APIs.
"""

from typing import Optional, Dict
from utils.http_client import fetch_json_sync

_meta_cache: Dict[str, dict] = {}


def resolve_meta(imdb_id: str, media_type: str) -> Optional[dict]:
    """Resolve title and year from IMDB/kitsu ID."""
    if not imdb_id:
        return None

    cache_key = f"{imdb_id}:{media_type}"
    if cache_key in _meta_cache:
        return _meta_cache[cache_key]

    result = None
    try:
        if imdb_id.startswith("kitsu:"):
            kid = imdb_id.replace("kitsu:", "")
            data = fetch_json_sync(f"https://kitsu.app/api/edge/anime/{kid}", timeout=4)
            if data and "data" in data:
                attrs = data["data"].get("attributes", {})
                sd = attrs.get("startDate", "")
                result = {
                    "name": attrs.get("canonicalTitle", ""),
                    "year": sd[:4] if sd else "",
                }
        else:
            ct = "movie" if media_type == "movie" else "series"
            data = fetch_json_sync(f"https://v3-cinemeta.strem.io/meta/{ct}/{imdb_id}.json", timeout=4)
            if data and "meta" in data:
                m = data["meta"]
                result = {
                    "name": m.get("name", ""),
                    "year": str(m.get("year", ""))[:4] if m.get("year") else "",
                }
    except Exception as e:
        print(f"[metadata] failed for {imdb_id}: {e}")

    if result:
        _meta_cache[cache_key] = result
    return result


def parse_stremio_id(raw_id: str) -> dict:
    """Parse a Stremio ID into components."""
    if not raw_id:
        return {"imdb_id": None, "kitsu_id": None, "season": None, "episode": None}

    clean = raw_id.replace(".json", "")

    for prefix in ["kitsu", "mal", "tvdb", "tmdb"]:
        if clean.startswith(f"{prefix}:"):
            parts = clean[len(prefix) + 1 :].split(":")
            return {
                "imdb_id": None,
                "kitsu_id": parts[0] if prefix == "kitsu" else None,
                "season": int(parts[1]) if len(parts) >= 3 and parts[1].isdigit() else None,
                "episode": int(parts[2]) if len(parts) >= 3 and parts[2].isdigit() else None,
            }

    parts = clean.split(":")
    return {
        "imdb_id": parts[0],
        "kitsu_id": None,
        "season": int(parts[1]) if len(parts) >= 3 and parts[1].isdigit() else None,
        "episode": int(parts[2]) if len(parts) >= 3 and parts[2].isdigit() else None,
    }

File: src/utils/metadata.js
// src/utils/metadata.js — Cinemeta + Kitsu resolver
import axios from 'axios';

const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36';
const cache = new Map();

export async function resolveMeta(imdbId, type) {
  if (!imdbId) return null;
  const key = `${imdbId}:${type}`;
  const cached = cache.get(key);
  if (cached && Date.now() - cached.ts < 1800000) return cached.data;

  let result = null;
  try {
    if (imdbId.startsWith('kitsu:')) {
      const id = imdbId.replace('kitsu:', '');
      const res = await axios.get(`https://kitsu.app/api/edge/anime/${id}`, { timeout: 4000, headers: { 'User-Agent': UA } });
      const a = res.data?.data?.attributes;
      if (a) result = { name: a.canonicalTitle || '', year: (a.startDate || '').slice(0, 4) };
    } else {
      const ct = type === 'movie' ? 'movie' : 'series';
      const res = await axios.get(`https://v3-cinemeta.strem.io/meta/${ct}/${imdbId}.json`, { timeout: 4000, headers: { 'User-Agent': UA } });
      const m = res.data?.meta;
      if (m) result = { name: m.name || '', year: m.year ? String(m.year).slice(0, 4) : '' };
    }
  } catch {}
  if (result) cache.set(key, { data: result, ts: Date.now() });
  return result;
}

export async function resolveTmdbId(imdbId, type, title) {
  const TMDB_KEY = '6b2dec73b6697866a50cdaef60ccffcb';
  const kind = type === 'movie' ? 'movie' : 'tv';
  try {
    if (imdbId && !imdbId.startsWith('kitsu:') && !imdbId.startsWith('tmdb:')) {
      const res = await axios.get(`https://api.themoviedb.org/3/find/${imdbId}?api_key=${TMDB_KEY}&external_source=imdb_id`, { timeout: 5000 });
      const arr = res.data?.[`${kind}_results`] || [];
      if (arr[0]?.id) return String(arr[0].id);
    }
    if (title) {
      const clean = title.replace(/\s+\d{4}$/, '').trim();
      const res = await axios.get(`https://api.themoviedb.org/3/search/${kind}?api_key=${TMDB_KEY}&query=${encodeURIComponent(clean)}`, { timeout: 5000 });
      if (res.data?.results?.[0]?.id) return String(res.data.results[0].id);
    }
  } catch {}
  return null;
}

export function parseStremioId(id) {
  if (!id) return { imdbId: null, kitsuId: null, season: null, episode: null };
  const raw = String(id).replace(/\.json$/, '');
  for (const p of ['kitsu', 'mal', 'tvdb', 'tmdb']) {
    if (raw.startsWith(p + ':')) {
      const parts = raw.slice(p.length + 1).split(':');
      return {
        imdbId: null,
        kitsuId: p === 'kitsu' ? parts[0] : null,
        season: parts.length >= 3 && /^\d+$/.test(parts[1]) ? Number(parts[1]) : null,
        episode: parts.length >= 3 && /^\d+$/.test(parts[2]) ? Number(parts[2]) : null,
      };
    }
  }
  const parts = raw.split(':');
  return {
    imdbId: parts[0],
    kitsuId: null,
    season: parts.length >= 3 && /^\d+$/.test(parts[1]) ? Number(parts[1]) : null,
    episode: parts.length >= 3 && /^\d+$/.test(parts[2]) ? Number(parts[2]) : null,
  };
}
